compiler/rustc_hir_analysis/src/lib.rs

```diff
@@ -204,7 +205,12 @@ pub fn check_crate(tcx: TyCtxt<'_>) -> Result<(), ErrorGuaranteed> {
         let def_kind = tcx.def_kind(item_def_id);
         match def_kind {
             DefKind::Static(_) => tcx.ensure().eval_static_initializer(item_def_id),
-            DefKind::Const => tcx.ensure().const_eval_poly(item_def_id.into()),
+            DefKind::Const if tcx.generics_of(item_def_id).params.is_empty() => {
```

Comment:
this impacts behavior for consts with where-bounds, do we get a similar perf improvement if we continue using `const_eval_poly` but change it to use `self.tcx.param_env_with_reveal_all(def_id)`?

Reply:
this is only an issue with `trivial_bounds`, which are checked on MIR in mir_drops_elaborated_and_const_checked. If they don't hold, we destroy the MIR body which will fail const eval. Not ideal I guess. I'll ponder on how to get perf and no issues at the same time 😆

Reply:
Hmm... I guess we already fail it today:

```rust
#![feature(trivial_bounds)]
#![feature(generic_const_items)]
const FOO: () = () where str: Sized;
```

with exactly the error I assumed

```
error[E0080]: evaluation of constant value failed
 --> src/lib.rs:3:1
  |
3 | const FOO: () = () where str: Sized;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ entering unreachable code
```

Reply:
how does this work with
```rust
trait Trait {
    const ASSOC: u32;
}

impl Trait for ()
where
    for<'a> &'a (): Trait,
{
    const ASSOC: u32 = <&'static ()>::ASSOC;
}
```

Reply:
That's not a `DefKind::Const`, but a `DefKind::AssocConst`

Reply:
does
```rust
#![feature(generic_const_items)]
trait Trait {
    const ASSOC: u32;
}

const ASSOC: u32 = <&'static ()>::ASSOC where for<'a> &'a (): Trait;
```
go from pass to error with this PR?

Reply:
nope, it still passes